### PR TITLE
feat(gitlab-release): implement datasource

### LINKED ISSUE #7534: feat(gitlab): gitlab-releases
**What would you like Renovate to be able to do?**
Would be nice to have also gitlab-releases datasource

**Did you already have any implementation ideas?**
Should be similar to this:
- https://github.com/renovatebot/renovate/tree/master/lib/datasource/gitlab-tags
- https://github.com/renovatebot/renovate/tree/master/lib/datasource/github-releases

**Are there any workarounds or alternative ideas you've tried to avoid needing this feature?**
no

**Is this a feature you'd be interested in implementing yourself?**
no

Docs: https://docs.gitlab.com/ee/api/releases/

### PR DESCRIPTION
**Pull request by secustor**
## Changes:
Implements a Gitlab Release datasource 

## Context:
Supersedes: https://github.com/renovatebot/renovate/pull/10626    
Closes https://github.com/renovatebot/renovate/issues/7534

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by rarkins — DISMISSED** [review]

File: lib/datasource/gitlab-releases/index.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  async getReleases({
+    registryUrl,
+    lookupName,
+  }: GetReleasesConfig): Promise<ReleaseResult | null> {
+    const urlEncodedRepo = encodeURIComponent(lookupName);
+    const apiUrl = `${registryUrl}/api/v4/projects/${urlEncodedRepo}/releases`;
+
+    const gitlabReleasesResponse = (
+      await this.http.getJson<GitlabRelease[]>(apiUrl)
+    ).body;
```
Comment (JamieMagee) [inline lib/datasource/gitlab-releases/index.ts:31]: Can you add some error handling here? `handleGenericErrors` will handle 429s and 5xxs for you.

File: lib/datasource/gitlab-releases/index.spec.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        .get('/api/v4/projects/some%2Fdep2/releases')
+        .reply(200, body);
+      const res = await getPkgReleases({
+        datasource: GitlabReleasesDatasource.id,
+        registryUrls: ['https://gitlab.company.com'],
+        depName: 'some/dep2',
+      });
+      expect(res).toMatchSnapshot();
+      expect(res.releases).toHaveLength(2);
+      expect(httpMock.getTrace()).toMatchSnapshot();
```
Comment (viceice) [inline lib/datasource/gitlab-releases/index.spec.ts:31]:
```suggestion
```

Not needed, will be removed from other tests soon

File: lib/datasource/gitlab-releases/index.spec.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        .scope('https://gitlab.com')
+        .get('/api/v4/projects/some%2Fdep2/releases')
+        .reply(200, body);
+      const res = await getPkgReleases({
+        datasource: GitlabReleasesDatasource.id,
+        depName: 'some/dep2',
+      });
+      expect(res).toMatchSnapshot();
+      expect(res.releases).toHaveLength(2);
+      expect(httpMock.getTrace()).toMatchSnapshot();
```
Comment (viceice) [inline lib/datasource/gitlab-releases/index.spec.ts:45]:
```suggestion
```

File: lib/datasource/gitlab-releases/index.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+export class GitlabReleasesDatasource extends Datasource {
+  static readonly id = 'gitlab-releases';
+
+  readonly defaultRegistryUrls = ['https://gitlab.com'];
+
+  static readonly registryStrategy = 'first';
+
+  constructor() {
+    super(GitlabReleasesDatasource.id);
```
Comment (viceice) [inline lib/datasource/gitlab-releases/index.ts:14]: I think we also need to override `Http` by `GitlabHttp` class, so we reuse buildin paging.

File: lib/datasource/gitlab-releases/types.ts [hunk 5]
```diff
@@ -0,0 +1,5 @@
+export type GitlabRelease = {
```
Comment (viceice) [inline lib/datasource/gitlab-releases/types.ts:1]:
```suggestion
export interface GitlabRelease {
```

Prefer interfaces

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/datasource/gitlab-releases/index.ts [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+import type { GitlabRelease } from './types';
+
+export class GitlabReleasesDatasource extends Datasource {
+  static readonly id = 'gitlab-releases';
+
+  readonly defaultRegistryUrls = ['https://gitlab.com'];
+
+  static readonly registryStrategy = 'first';
+
+  http = new GitlabHttp();
```
Comment (viceice) [inline lib/datasource/gitlab-releases/index.ts:14]: I think we need to add a new constructor to override the hostType, see #11136 for github sample

**Review by viceice — CHANGES_REQUESTED** [review]
needs code coverage

File: lib/types/platform/gitlab/index.ts [hunk 7]
```diff
@@ -12,3 +13,10 @@ export type GitlabTreeNode = {
   id: string;
   name: string;
 } & GitTreeNode;
+
+export const GITLAB_API_USING_HOST_TYPES = [
+  PLATFORM_TYPE_GITLAB,
+  'gitlab-releases',
+  'gitlab-tags',
+  'pod',
+];
```
Comment (viceice) [inline lib/types/platform/gitlab/index.ts:22]: Can you add verification tests like in https://github.com/renovatebot/renovate/blob/main/lib/types/github.spec.ts

File: lib/util/http/gitlab.spec.ts [hunk 8]
```diff
@@ -68,6 +71,21 @@ describe('util/http/gitlab', () => {
...
   });
+
+  it('supports different datasources', async () => {
+    const gitlabApiDatasource = new GitlabHttp(GitlabReleasesDatasource.id);
+    hostRules.add({ hostType: PLATFORM_TYPE_GITLAB, token: 'abc' });
+    hostRules.add({
+      hostType: GitlabReleasesDatasource.id,
+      token: 'def',
+    });
+    httpMock.scope(gitlabApiHost).get('/api/v4/some-url').reply(200);
```
Comment (viceice) [inline lib/util/http/gitlab.spec.ts:82]:
```suggestion
    httpMock.scope(gitlabApiHost, { reqheaders: {authorization: 'Bearer def'}}).get('/api/v4/some-url').reply(200);
```

use nock header matching

https://github.com/nock/nock#specifying-headers

File: lib/util/http/gitlab.spec.ts [hunk 9]
```diff
@@ -68,6 +71,21 @@ describe('util/http/gitlab', () => {
...
+    await gitlabApiDatasource.get('/some-url');
+    const [req] = httpMock.getTrace();
+    expect(req).toBeDefined();
+    expect(req.headers.authorization).toBe('Bearer def');
```
Comment (viceice) [inline lib/util/http/gitlab.spec.ts:86]:
```suggestion
```

See above

**Review by viceice — COMMENTED** [review]
Some small test stuff, otherwise LGTM

**Comment by secustor** [comment]
Not sure if you can simply accept the issue or I have to do something on my part.      
I couldn't find something in the lines of `ignore the next line`

File: lib/types/platform/gitlab/index.ts [hunk 10]
```diff
@@ -12,3 +13,10 @@ export type GitlabTreeNode = {
   id: string;
   name: string;
 } & GitTreeNode;
+
+export const GITLAB_API_USING_HOST_TYPES = [
+  PLATFORM_TYPE_GITLAB,
+  'gitlab-releases',
+  'gitlab-tags',
+  'pod',
```
Comment (rarkins) [inline lib/types/platform/gitlab/index.ts:21]: Won't this now "hijack" all pod to use gitlab auth?

Comment (viceice) [reply lib/types/platform/gitlab/index.ts:21]:
```suggestion
```

Yes i think so, not sure why it's added. So suggest to remove that

File: lib/types/platform/gitlab/index.spec.ts [hunk 11]
```diff
@@ -0,0 +1,27 @@
+import {
+  PLATFORM_TYPE_GITHUB,
+  PLATFORM_TYPE_GITLAB,
+} from '../../../constants/platforms';
+import { GitlabReleasesDatasource } from '../../../datasource/gitlab-releases';
+import { id as GL_TAGS_DS } from '../../../datasource/gitlab-tags';
+import { id as POD_DS } from '../../../datasource/pod';
```
Comment (viceice) [inline lib/types/platform/gitlab/index.spec.ts:7]:
```suggestion
```

File: lib/types/platform/gitlab/index.spec.ts [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+import { GITLAB_API_USING_HOST_TYPES } from './index';
+
+describe('types/platform/gitlab/index', () => {
+  it('should be part of the GITLAB_API_USING_HOST_TYPES', () => {
+    expect(GITLAB_API_USING_HOST_TYPES.includes(GL_TAGS_DS)).toBeTrue();
+    expect(
+      GITLAB_API_USING_HOST_TYPES.includes(GitlabReleasesDatasource.id)
+    ).toBeTrue();
+    expect(GITLAB_API_USING_HOST_TYPES.includes(POD_DS)).toBeTrue();
```
Comment (viceice) [inline lib/types/platform/gitlab/index.spec.ts:16]:
```suggestion
```

**Review by viceice — COMMENTED** [review]
POD datasource is either using github or official / generic cdn

File: lib/types/platform/gitlab/index.ts (hunk 10 above)
Comment (secustor) [reply lib/types/platform/gitlab/index.ts]: Ah sry, I simply adapted the Github plattform example and haven't realized that `pod` is referencing a datasource. My Impressions has been that that has something to do that the Renovate App is probably deployed inside of a K8 cluster.

**Review by viceice — APPROVED** [review]

**Pull request merged**